[lldb][Mach-O] Allow "process metadata" LC_NOTE to supply registers

jasonmolenda opened the pull request:
The "process metadata" LC_NOTE allows for thread IDs to be specified in a Mach-O corefile.  This extends the JSON recognzied in that LC_NOTE to allow for additional registers to be supplied on a per-thread basis.

The registers included in a Mach-O corefile LC_THREAD load command can only be one of the register flavors that the kernel (xnu) defines in <mach/arm/thread_status.h> for arm64 -- the general purpose registers, floating point registers, exception registers.

JTAG style corefile producers may have access to many additional registers beyond these that EL0 programs typically use, for instance TCR_EL1 on AArch64, and people developing low level code need access to these registers.  This patch defines a format for including these registers for any thread.

The JSON in "process metadata" is a dictionary that must have a `threads` key.  The value is an array of entries, one per LC_THREAD in the Mach-O corefile.  The number of entries must match the LC_THREADs so they can be correctly associated.

Each thread's dictionary must have two keys, `sets`, and `registers`. `sets` is an array of register set names.  If a register set name matches one from the LC_THREAD core registers, any registers that are defined will be added to that register set.  e.g. metadata can add a register to the "General Purpose Registers" set that lldb shows users.

`registers` is an array of dictionaries, one per register.  Each register must have the keys `name`, `value`, `bitsize`, and `set`. It may provide additional keys like `alt-name`, that `DynamicRegisterInfo::SetRegisterInfo` recognizes.

This `sets` + `registers` formatting is the same that is used by the `target.process.python-os-plugin-path` script interface uses, both are parsed by `DynamicRegisterInfo`.  The one addition is that in this LC_NOTE metadata, each register must also have a `value` field, with the value provided in big-endian base 10, as usual with JSON.

In RegisterContextUnifiedCore, I combine the register sets & registers from the LC_THREAD for a specific thread, and the metadata sets & registers for that thread from the LC_NOTE.  Even if no LC_NOTE is present, this class ingests the LC_THREAD register contexts and reformats it to its internal stores before returning itself as the RegisterContex, instead of shortcutting and returning the core's native RegisterContext.  I could have gone either way with that, but in the end I decided if the code is correct, we should live on it always.

I added a test where we process save-core to create a userland corefile, then use a utility "add-lcnote" to strip the existing "process metadata" LC_NOTE that lldb put in it, and adds a new one from a JSON string.

rdar://74358787

jasonmolenda committed:
[lldb][Mach-O] Allow "process metadata" LC_NOTE to supply registers

The "process metadata" LC_NOTE allows for thread IDs to be specified
in a Mach-O corefile.  This extends the JSON recognzied in that
LC_NOTE to allow for additional registers to be supplied on a
per-thread basis.

The registers included in a Mach-O corefile LC_THREAD load command
can only be one of the register flavors that the kernel (xnu) defines
in <mach/arm/thread_status.h> for arm64 -- the general purpose
registers, floating point registers, exception registers.

JTAG style corefile producers may have access to many additional
registers beyond these that EL0 programs typically use, for instance
TCR_EL1 on AArch64, and people developing low level code need access
to these registers.  This patch defines a format for including these
registers for any thread.

The JSON in "process metadata" is a dictionary that must have a
`threads` key.  The value is an array of entries, one per LC_THREAD
in the Mach-O corefile.  The number of entries must match the
LC_THREADs so they can be correctly associated.

Each thread's dictionary must have two keys, `sets`, and `registers`.
`sets` is an array of register set names.  If a register set name
matches one from the LC_THREAD core registers, any registers that
are defined will be added to that register set.  e.g. metadata can
add a register to the "General Purpose Registers" set that lldb
shows users.

`registers` is an array of dictionaries, one per register.  Each
register must have the keys `name`, `value`, `bitsize`, and `set`.
It may provide additional keys like `alt-name`, that
`DynamicRegisterInfo::SetRegisterInfo` recognizes.

This `sets` + `registers` formatting is the same that is used by
the `target.process.python-os-plugin-path` script interface uses,
both are parsed by `DynamicRegisterInfo`.  The one addition is that
in this LC_NOTE metadata, each register must also have a `value`
field, with the value provided in big-endian base 10, as usual with
JSON.

In RegisterContextUnifiedCore, I combine the register sets & registers
from the LC_THREAD for a specific thread, and the metadata sets &
registers for that thread from the LC_NOTE.  Even if no LC_NOTE
is present, this class ingests the LC_THREAD register contexts and
reformats it to its internal stores before returning itself as the
RegisterContex, instead of shortcutting and returning the core's
native RegisterContext.  I could have gone either way with that,
but in the end I decided if the code is correct, we should live on
it always.

I added a test where we process save-core to create a userland corefile,
then use a utility "add-lcnote" to strip the existing "process metadata"
LC_NOTE that lldb put in it, and adds a new one from a JSON string.

rdar://74358787

jasonmolenda commented:
The API test case included in the PR makes it pretty easy to see what the format looks like, if the english description wasn't clear.

jasonmolenda committed:
remove debug print

jasonmolenda committed:
appease clang format

JDevlieghere commented on lldb/source/Plugins/Process/mach-core/RegisterContextUnifiedCore.cpp:
```diff
@@ -0,0 +1,293 @@
+  // If we have thread metadata, check if the keys for register
+  // definitions are present; if not, clear the ObjectSP.
+  if (metadata_thread_registers &&
+      metadata_thread_registers->GetAsDictionary()->HasKey("register_info")) {
```
This might crash if `metadata_thread_registers` is not a dict. You need to do something like:

```
if (metadata_thread_registers) {
  if (auto* dict = metadata_thread_registers->GetAsDictionary()) {
    if (dict->->HasKey("register_info"))
      metadata_registers_dict = dict->GetValueForKey("register_info")->GetAsDictionary();
  }
}

```

JDevlieghere commented on lldb/source/Plugins/Process/mach-core/RegisterContextUnifiedCore.cpp:
```diff
@@ -0,0 +1,293 @@
+  }
+
+  // Create a DynamicRegisterInfo from the metadata JSON.
+  std::unique_ptr<DynamicRegisterInfo> additional_reginfo_up;
+  if (metadata_registers_dict)
+    additional_reginfo_up = DynamicRegisterInfo::Create(
+        *metadata_registers_dict, target.GetArchitecture());
+
```
```suggestion
    std::unique_ptr<DynamicRegisterInfo>
        additional_reginfo_up =
            metadata_registers_dict
                ? DynamicRegisterInfo::Create(*metadata_registers_dict,
                                              target.GetArchitecture())
                : nullptr;
```

JDevlieghere commented on lldb/source/Plugins/Process/mach-core/RegisterContextUnifiedCore.cpp:
```diff
@@ -0,0 +1,293 @@
+    RegisterSet new_set;
+    const RegisterSet *old_set = core_thread_regctx_sp->GetRegisterSet(idx);
+    new_set.name = ConstString(old_set->name).AsCString();
+    if (old_set->short_name)
+      new_set.short_name = ConstString(old_set->short_name).AsCString();
+    else
+      new_set.short_name = nullptr;
```
```suggestion
    new_set.short_name = old_set->short_name ? ConstString(old_set->short_name).AsCString() : nullptr;
```

JDevlieghere commented on lldb/source/Plugins/Process/mach-core/RegisterContextUnifiedCore.cpp:
```diff
@@ -0,0 +1,293 @@
+    lldb_private::RegisterValue &value) {
+  if (!reg_info)
+    return false;
+  ProcessSP process_sp(m_thread.GetProcess());
+  if (process_sp) {
```
```suggestion
  if (ProcessSP process_sp = m_thread.GetProcess()) {
```

JDevlieghere commented on lldb/source/Plugins/Process/mach-core/RegisterContextUnifiedCore.h:
```diff
@@ -0,0 +1,57 @@
+//===-- RegisterContextUnifiedCore.h ----------------------------*- C++ -*-===//
```
```suggestion
//===-- RegisterContextUnifiedCore.h --------------------------------------===//
```

JDevlieghere commented on lldb/source/Plugins/Process/mach-core/RegisterContextUnifiedCore.h:
```diff
@@ -0,0 +1,57 @@
+private:
+  std::vector<lldb_private::RegisterSet> m_register_sets;
+  std::vector<lldb_private::RegisterInfo> m_register_infos;
+  // For each register set, an array of register numbers included.
```
```suggestion
  /// For each register set, an array of register numbers included.
```

JDevlieghere commented on lldb/source/Plugins/Process/mach-core/RegisterContextUnifiedCore.h:
```diff
@@ -0,0 +1,57 @@
+  std::vector<lldb_private::RegisterInfo> m_register_infos;
+  // For each register set, an array of register numbers included.
+  std::map<size_t, std::vector<uint32_t>> m_regset_regnum_collection;
+  // Bytes of the register contents.
```
```suggestion
  /// Bytes of the register contents.
```

JDevlieghere commented on lldb/source/Plugins/Process/mach-core/RegisterContextUnifiedCore.cpp:
```diff
@@ -0,0 +1,293 @@
+  // Set up our RegisterSet array.
+  // First copying all of the core thread register sets,
+  // then any additional unique register sets from the metadata.
+  if (additional_reginfo_up) {
+    for (size_t idx = 0; idx < additional_reginfo_up->GetNumRegisterSets();
+         idx++) {
+      bool found_match = false;
+      const RegisterSet *old_set = additional_reginfo_up->GetRegisterSet(idx);
+      for (size_t jdx = 0; jdx < m_register_sets.size(); jdx++) {
+        if (strcmp(m_register_sets[jdx].name, old_set->name) == 0) {
+          metadata_regset_to_combined_regset[idx] = jdx;
+          found_match = true;
+          break;
+        }
+      }
+      if (!found_match) {
+        RegisterSet new_set;
+        new_set.name = ConstString(old_set->name).AsCString();
+        if (old_set->short_name)
+          new_set.short_name = ConstString(old_set->short_name).AsCString();
+        else
+          new_set.short_name = nullptr;
+        metadata_regset_to_combined_regset[idx] = m_register_sets.size();
+        m_register_sets.push_back(new_set);
+      }
+    }
+  }
```
This seems almost (?) identical to the code above? Can this be extracted into a helper that's passed `metadata_registers_dict` and `additional_reginfo_up` respectively?

jasonmolenda committed:
Update lldb/source/Plugins/ObjectFile/Mach-O/ObjectFileMachO.cpp

Co-authored-by: Jonas Devlieghere <jonas@devlieghere.com>

jasonmolenda committed:
ws fix

jasonmolenda committed:
Update lldb/source/Plugins/Process/mach-core/RegisterContextUnifiedCore.cpp

Co-authored-by: Jonas Devlieghere <jonas@devlieghere.com>

jasonmolenda committed:
Update lldb/source/Plugins/Process/mach-core/RegisterContextUnifiedCore.h

Co-authored-by: Jonas Devlieghere <jonas@devlieghere.com>

jasonmolenda committed:
Update lldb/source/Plugins/Process/mach-core/RegisterContextUnifiedCore.h

Co-authored-by: Jonas Devlieghere <jonas@devlieghere.com>

jasonmolenda committed:
Update lldb/source/Plugins/Process/mach-core/RegisterContextUnifiedCore.h

Co-authored-by: Jonas Devlieghere <jonas@devlieghere.com>

jasonmolenda committed:
Merge branch 'main' into lc_note-metadata-register-additions

jasonmolenda committed:
Add two tests with invalid process metadata in the corefile.
Correct lldb to reject both invalid process metadata
corefiles.

jasonmolenda committed:
Check that the per-thread JSON is a dictionary
before treating it as such in RegisterContextUnifiedCore.

jasonmolenda committed:
Do a pass over RegisterContextUnifiedCore::RegisterContextUnifiedCore
simplifying the code a bit, making the comments clearer.

jasonmolenda commented:
@JDevlieghere I think I'm done fiddling with this.  I thought about breaking out the block of code that sets up the combined register sets, but I don't think it makes things much clearer.  I did use a lambda to copy the register set names into the constant string pool when creating our new one, so we don't get burned by char* lifetime issues.

JDevlieghere commented on lldb/source/Plugins/ObjectFile/Mach-O/ObjectFileMachO.cpp:
```diff
@@ -5857,6 +5838,50 @@ bool ObjectFileMachO::GetCorefileThreadExtraInfos(
+        log,
+        "Multiple 'process metadata' LC_NOTEs found, only using the first.");
+
+  offset_t payload_offset = std::get<0>(lc_notes[0]);
+  offset_t strsize = std::get<1>(lc_notes[0]);
```
FWIW, with structured bindings in C++17 you can now write this like:
```suggestion
  auto [payload_offset, strsize] = lc_notes[0];
```

JDevlieghere commented on lldb/source/Plugins/ObjectFile/Mach-O/ObjectFileMachO.cpp:
```diff
@@ -5857,6 +5838,50 @@ bool ObjectFileMachO::GetCorefileThreadExtraInfos(
+    buf.resize(buf.size() - 1);
+  StructuredData::ObjectSP object_sp = StructuredData::ParseJSON(buf);
+  if (!object_sp) {
+    LLDB_LOGF(log, "Unable to read 'process metadata' LC_NOTE, did not "
+                   "parse as valid JSON.");
```
No need to change anything, but another way to handle this is to make this function return an `Expected<StructuredData::ObjectSP>` and do the logging on the caller side. The advantage of that is that if you can do better than logging, it's easy to percolate the error up, while now it never leaves the function.

JDevlieghere commented on lldb/source/Plugins/ObjectFile/Mach-O/ObjectFileMachO.h:
```diff
@@ -133,6 +133,8 @@ class ObjectFileMachO : public lldb_private::ObjectFile {
 
   bool GetCorefileThreadExtraInfos(std::vector<lldb::tid_t> &tids) override;
 
+  lldb_private::StructuredData::ObjectSP GetCorefileProcessMetadata() override;
```
```suggestion
 StructuredData::ObjectSP GetCorefileProcessMetadata() override;
```

JDevlieghere commented on lldb/source/Plugins/Process/mach-core/RegisterContextUnifiedCore.cpp:
```diff
@@ -0,0 +1,300 @@
+             val.GetByteSize());
+  }
+
+  // Set 'offset' fields for each register definition into our combined
+  // register data buffer. DynamicRegisterInfo needs
+  // this field set to parse the JSON.
```
Nit: looks like this needs to be reflowed.

JDevlieghere commented on lldb/source/Plugins/Process/mach-core/RegisterContextUnifiedCore.cpp:
```diff
@@ -0,0 +1,300 @@
+                                                             bitsize))
+          return false;
+        if (!ent->GetAsDictionary()->GetValueForKeyAsInteger("value", value)) {
+          // we had a bitsize but no value, so move the offset forward I guess.
```
```suggestion
          // We had a bitsize but no value, so move the offset forward I guess.
```

JDevlieghere commented on lldb/source/Plugins/Process/mach-core/RegisterContextUnifiedCore.cpp:
```diff
@@ -0,0 +1,300 @@
+        }
+        ent->GetAsDictionary()->AddIntegerItem("offset", offset);
+        Status error;
+        int bytesize = bitsize / 8;
```
Can this be const to make it clear that this doesn't change? 
```suggestion
        const int bytesize = bitsize / 8;
```

JDevlieghere commented on lldb/source/Plugins/Process/mach-core/RegisterContextUnifiedCore.cpp:
```diff
@@ -0,0 +1,300 @@
+        if (metadata_regset_to_combined_regset[setidx] == combined_regset_idx) {
+          const RegisterSet *regset =
+              additional_reginfo_up->GetRegisterSet(setidx);
+          for (size_t j = 0; j < regset->num_registers; j++) {
+            uint32_t reg_idx = regset->registers[j];
+            const RegisterInfo *reginfo =
+                additional_reginfo_up->GetRegisterInfoAtIndex(reg_idx);
+            m_regset_regnum_collection[combined_regset_idx].push_back(
+                m_register_infos.size());
+            // register names in DynamicRegisterInfo are ConstString stored;
+            // we can reuse the char* pointers here without retaining the
+            // DynamicRegisterInfo.
+            m_register_infos.push_back(*reginfo);
+            registers_this_regset++;
+          }
```
Isn't this code identical to the codeon line 212? Can you extract a helper/lambda that takes a `const RegisterSet *`?

jasonmolenda replied:
Yeah lol, there's like 3 lines I could pull out of these two loops into a lambda.

jasonmolenda replied:
yeah punting on this, it's more lines overall to put those in a lambda.

JDevlieghere commented on lldb/source/Plugins/Process/mach-core/RegisterContextUnifiedCore.cpp:
```diff
@@ -0,0 +1,300 @@
+    lldb_private::RegisterValue &value) {
+  if (!reg_info)
+    return false;
+  if (ProcessSP process_sp = m_thread.GetProcess()) {
```
I think I suggested this, but on second thought, if you early return you can save a whole level of indentation and the reader doesn't need to count to confirm whether this returns true or false on line 291 or 293 when process_sp is null.

JDevlieghere commented on lldb/source/Plugins/Process/mach-core/ThreadMachCore.cpp:
```diff
@@ -70,27 +77,49 @@ lldb::RegisterContextSP ThreadMachCore::GetRegisterContext() {
-  if (concrete_frame_idx == 0) {
-    if (!m_thread_reg_ctx_sp) {
-      ProcessSP process_sp(GetProcess());
+  ProcessSP process_sp(GetProcess());
```
The old code didn't check either, but let's assert that you must have a process at this point.
```suggestion
  ProcessSP process_sp(GetProcess());
  assert(process_sp);
```

jasonmolenda replied:
I mean, a Thread that wasn't constructed with a Process, you've got Problems. :)

JDevlieghere reviewed (commented):
Mostly nits

jasonmolenda committed:
Update lldb/source/Plugins/ObjectFile/Mach-O/ObjectFileMachO.cpp

Co-authored-by: Jonas Devlieghere <jonas@devlieghere.com>

jasonmolenda committed:
Update lldb/source/Plugins/Process/mach-core/RegisterContextUnifiedCore.cpp

Co-authored-by: Jonas Devlieghere <jonas@devlieghere.com>

jasonmolenda committed:
Update lldb/source/Plugins/Process/mach-core/ThreadMachCore.cpp

Co-authored-by: Jonas Devlieghere <jonas@devlieghere.com>

jasonmolenda committed:
Update lldb/source/Plugins/Process/mach-core/RegisterContextUnifiedCore.cpp

Co-authored-by: Jonas Devlieghere <jonas@devlieghere.com>

jasonmolenda committed:
reflow comment

jasonmolenda committed:
Update lldb/source/Plugins/ObjectFile/Mach-O/ObjectFileMachO.h

Co-authored-by: Jonas Devlieghere <jonas@devlieghere.com>

jasonmolenda committed:
Take another pass at the comments where the
combined RegisterInfo array is constructed from
the two different sources of register information.
